content/architecture/nanoclr-stub-args.md

```diff
@@ -0,0 +1,209 @@
+```
+
+So they allow you to setup the return element and leave. Those are so far used only in the `src\CLR\Core\TypeSystem.cpp` file. That said, nothing prevent you to use them as well.
+
+## CLR_RT_StackFrame &stack
+
+At every call to one of those native functions, the stack is passed thru a structure called CLR_RT_StackFrame. The definition can be found in `src\CLR\Include\nanoCLR_Runtime.h`. It would be too long to go thru all the functions and properties. In this description we will focus only on few elements.
+
+### If your function is in a static class
+
+In, this case you **won't** have the class stack available. Stack class is only available for non static class.
+
+### Getting and checking the stack in a non static class
+
+The pattern to use is the following:
+
+```cpp
+CLR_RT_HeapBlock* pThis = stack.This();
+FAULT_ON_NULL(pThis);
+```
+
+`FAULT_ON_NULL` is a macro similar to the onces presented in the previous section which will check of the stack is null or not. In case of null it will go to the `nanoCLR_Cleanup` label and set the `HRESULT` to a null fault.
+
+The definition can be found in `src\CLR\Include\nanoCLR_Interop.h`.
+
+```cpp
+#define FAULT_ON_NULL(ptr)							if(!(ptr)) NANOCLR_SET_AND_LEAVE(CLR_E_NULL_REFERENCE)
+#define FAULT_ON_NULL_ARG(ptr)						if(!(ptr)) NANOCLR_SET_AND_LEAVE(CLR_E_ARGUMENT_NULL)
+```
+
+You can use those macro for arguments too. We will see this in one of the following section.
+
+### Getting any exposed field from the stack
+
+Once the stack is valid, you can get any exposed field. Here is a typical example:
+
+```cpp
+int pinNumber = (int)(pThis[Library_sys_dev_pwm_native_System_Device_Pwm_PwmChannel::FIELD___pinNumber].NumericByRef().u4);
+```
+
+The stack is a `CLR_RT_HeapBlock`. This type is the core type that allows you get access to the stack.
+
+The pattern to use is the array one: `pthis[the_field_to_get]` where you have to make sure the field does exist. To avoid any issue, it is recommended to use the long names like `Library_sys_dev_pwm_native_System_Device_Pwm_PwmChannel::FIELD___pinNumber`.
+
+Then the `NumericByRef()` allows you to get a numerical number and `u4` to convert it as a `uint32`. For numerics, you have as well `NumericByRefConst()` allowing to convert as a constant. Then you have `u` for non signed, 1, 2, 4 and 8 for the number of bytes. `i` for signed, `r4` for `float` and `r8` for `double`.
+
+> Pro tip: the generated function definition contains the types and the return type. Example: `NativeSetActiveDutyCyclePercentage___VOID__U4` is void as return and U4 so uint32 for the first argument. This is done on purpose to help managing the return type and arguments.
+
+## Arg0, Arg1, ArgsN
+
+Helpers functions available to access the arguments. We will look at this in this section.
+
+### Static class
+
+In a static class, the `stack.Arg0()` represent the first argument. As explained before do **not** use `CLR_RT_HeapBlock* pThis = stack.This();` as this is not defined.
```

Comment:
```suggestion
In a static class, the `stack.Arg0()` points the first Heap Block passed on the IL stack. As explained before, do **not** use `CLR_RT_HeapBlock* pThis = stack.This();` because that won't be a pointer to a class instance (again: this is a call to a static method, therefore there is no instance of the class).
```